src/hooks/useWithdrawTokens.ts

```diff
@@ -0,0 +1,72 @@
+import { useState, useEffect } from 'react'
+import { TxResult, TokenBalanceDetails, TxOptionalParams } from 'types'
+import assert from 'assert'
+import { depositApi, walletApi } from 'api'
+import { ZERO } from 'const'
+
+interface Params {
+  tokenBalances: TokenBalanceDetails
+  txOptionalParams?: TxOptionalParams
+}
+
+interface Result {
+  withdrawable: boolean
+  withdrawing: boolean
+  withdraw(): Promise<TxResult<void>>
+  error: boolean
+}
+
+export const useWithdrawTokens = (params: Params): Result => {
+  const {
+    tokenBalances: { enabled, address: tokenAddress, withdrawingBalance },
+  } = params
+  const [withdrawable, setWithdrawable] = useState(false)
+  const [withdrawing, setWithdrawing] = useState(false)
+  const [error, setError] = useState(false)
+
+  async function _checkWithdrawable(): Promise<boolean> {
+    if (!withdrawingBalance.gt(ZERO)) {
+      return false
+    }
+    // TODO: Remove connect once login is done
+    await walletApi.connect()
+
+    // TODO: these two (useAddress and batchId) are constant for a given withdraw
+    //       probably could be cached?
+    const userAddress = await walletApi.getAddress()
+    const [withdrawBatchId, currentBatchId] = await Promise.all([
+      depositApi.getPendingWithdrawBatchId(userAddress, tokenAddress),
+      depositApi.getCurrentBatchId(),
+    ])
+
+    return withdrawBatchId < currentBatchId
+  }
+
+  useEffect(() => {
+    _checkWithdrawable()
+      .then(withdrawable => setWithdrawable(withdrawable))
+      .catch(error => {
+        console.error('Error checking withdraw state')
+        setError(error)
+      })
+  })
+
+  async function withdraw(): Promise<TxResult<void>> {
+    assert(enabled, 'Token not enabled')
+    assert(withdrawable, 'Withdraw not ready')
+
+    setWithdrawing(true)
+    console.debug(`Is withdrawing? ${withdrawing}`)
+    // TODO: Remove connect once login is done
+    await walletApi.connect()
+
+    const userAddress = await walletApi.getAddress()
+    const result = depositApi.withdraw(userAddress, tokenAddress, params.txOptionalParams)
+
+    setWithdrawing(false)
```

Comment:
Note that I made this mistake and had to fix it in the Memory Leak PR (#74)

Every time you have an await, things can happen in the middle. For example, the user can navigate, and the componet get's distroyed while executing line `64`.

Then, line `66` will do something that will trigger a redraw of the component. But if there's no component, we'll have a small issue.

mmmm, actually maybe this is the issue you where asking, I don't see the `await` keyword in line `64`

Reply:
That was it (facepalm)